File: 2016/puzzle_data.py
"""Implementation of PuzzleData class, which represents input data and
model solutions for each day's puzzle."""

# There are usually two sub puzzles using the same input. AOC website
# calls them part1 and part2 but we use a and b to distinguihs from
# the puzzle day number.
SOLUTIONS = ['a', 'b']


class PuzzleData:
    """Implements PuzzleData to store puzzle inputs and solutions.
    Some or all of this data may be missing."""
    def __init__(self, path):
        """Takes path to directory containing input and solution files."""
        self.input = []
        self.solutions = {}
        self.read_input(path)
        self.read_solutions(path)

    def read_input(self, path):
        """Read input data from file, ignoring errors."""
        file_name = path + "/input"
        try:
            with open(file_name, encoding='utf-8') as file:
                self.input = file.read().splitlines()
        except FileNotFoundError:
            pass

    def read_solutions(self, path):
        """Read solutions from file, ignoring errors."""
        base_file_name = path + "/solution."
        try:
            for suffix in SOLUTIONS:
                with open(base_file_name + suffix, encoding='utf-8') as file:
                    sol = file.read().splitlines()
                    self.solutions[suffix] = sol[0] if len(sol) == 1 else sol
        except FileNotFoundError:
            pass

    def has_input(self):
        """Return true if input data provided and not empty."""
        return len(self.input) >= 1

    def input_as_lines(self):
        """Return input as an array of strings."""
        return self.input

    def input_as_string(self):
        """Return input as a single string (first line)."""
        return self.input[0]

    def has_solution(self, key):
        """Returns true if a solution for key was provided."""
        assert key in SOLUTIONS
        return key in self.solutions

    def solution(self, key):
        """Returns a given solution."""
        assert key in SOLUTIONS
        return self.solutions[key]

File: 2017/aoc2017/source/puzzle_data.cpp
#include <fstream>
#include <string>
#include <vector>

#include "puzzle_data.hpp"

PuzzleData::PuzzleData(const std::string &path)
    : input_{read_input(path)}, solutions_{read_solutions(path)} {}

auto PuzzleData::has_input() const -> bool { return !input_.empty(); }

auto PuzzleData::input_as_lines() const -> const std::vector<std::string> & {
    return input_;
}

auto PuzzleData::input_as_string() const -> const std::string & {
    // Treat the input as a single line
    return input_[0];
}

auto PuzzleData::input_as_int() const -> int {
    // Treat the input as a single number
    return std::stoi(input_[0]);
}

auto PuzzleData::has_solution(SolutionPart::Type part) const -> bool {
    return solutions_.find(part) != solutions_.end();
}

auto PuzzleData::solution(SolutionPart::Type part) const -> std::string {
    return solutions_.at(part);
}

auto PuzzleData::read_input(const std::string &path) -> InputVector {
    PuzzleData::InputVector lines;
    std::ifstream file(path + "/input");
    if (file) {
        std::string line;
        while (std::getline(file, line)) {
            lines.push_back(line);
        }
        file.close();
    }
    return lines;
}

auto PuzzleData::read_solutions(const std::string &path) -> SolutionMap {
    PuzzleData::SolutionMap solutions;
    for (auto suffix : SolutionPart::All) {
        std::ifstream file(path + "/solution." + static_cast<char>(suffix));
        if (file) {
            std::string line;
            if (file >> line) {
                solutions[suffix] = line;
            }
            file.close();
        }
    }
    return solutions;
}
